modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,718 @@
+        }
+    }
+
+    /** Local gateway before global lock. */
+    private static final class LocalSync implements Lock {
+        /** */
+        private final ReentrantLock lock = new ReentrantLock();
+
+        /** */
+        private final AtomicLong threadCount = new AtomicLong();
+
+        /** */
+        private volatile boolean onGlobalLock = false;
+
+        /** */
+        private final GlobalUnfairSync globalSync;
+
+        /** */
+        private static final long MAX_TIME = 50_000_000L;
+
+        /** */
+        private volatile long nextFinish;
+
+        /** */
+        private final ThreadLocal<Boolean> hasLocked = new ThreadLocal<Boolean>() {
+            /** {@inheritDoc} */
+            @Override protected Boolean initialValue() {
+                return false;
+            }
+        };
+
+        /** */
+        private LocalSync(GlobalUnfairSync globalSync) {
+            assert globalSync != null;
+
+            this.globalSync = globalSync;
+        }
+
+        /** */
+        private boolean isLocked() throws IgniteCheckedException {
+            if (hasLocked.get())
+                return true;
+
+            if (onGlobalLock || lock.isLocked())
+                return true;
+
+            ArrayDeque<UUID> nodes = globalSync.forceGet().nodes;
+
+            return !(nodes == null || nodes.isEmpty());
+        }
+
+        /** */
+        private boolean holdGlobalLockInterruptibly() {
+            if (!onGlobalLock) {
+                if (!globalSync.acquireInterruptibly())
+                    return false;
+
+                onGlobalLock = true;
+
+                nextFinish = MAX_TIME + System.nanoTime();
+            }
+            return true;
+        }
+
+        /** */
+        private void releaseGlobalLock() {
+            if (threadCount.decrementAndGet() <= 0 || System.nanoTime() >= nextFinish) {
+                globalSync.release();
+
+                onGlobalLock = false;
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override public void lock() {
+            if (hasLocked.get())
+                return;
+
+            threadCount.incrementAndGet();
+
+            lock.lock();
+
+            if (!onGlobalLock) {
+                globalSync.acquire();
+
+                onGlobalLock = true;
+
+                nextFinish = MAX_TIME + System.nanoTime();
+            }
+
+            hasLocked.set(true);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void unlock() {
+            if (!hasLocked.get())
+                return;
+
+            try {
+                releaseGlobalLock();
+            }
+            finally {
+                lock.unlock();
+
+                hasLocked.set(false);
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override public void lockInterruptibly() throws InterruptedException {
+            if (hasLocked.get())
+                return;
+
+            threadCount.incrementAndGet();
```

Comment:
why haven't u implemneted all methods from java util concurrent locks ReentrantLock?
For instnce, getQueueLength.
When u implement getQueueLength() you may use threadCount variable. So, this variable must be incremented **atomically with lock.lockInterruptibly()**

Reply:
fixed